### ISSUE TITLE
[dereference.of.nullable] on conditional final field access

### ISSUE THREAD

**Issue by mtf90** [typetools/checker-framework#4872]
```java
import org.checkerframework.checker.nullness.qual.Nullable;

class CF {

    private final @Nullable Ref ref;

    CF(@Nullable Ref ref) {
        this.ref = ref;
    }

    void fun(CF other) {
        if (this.ref == null && other.ref == null) {
            System.out.println("Both null");
        } else if (this.ref == null) {
            System.out.println(other.ref.fun() + " should not be null");
        } else if (other.ref == null) {
            System.out.println(this.ref.fun() + " should not be null");
        } else {
            System.out.println(this.ref.fun() + " and " + other.ref.fun() + " should not be null");
        }
    }

    static class Ref {
        String fun() { return "fun"; }
    }
}

```

Running the default `NullnessChecker` 
 ```
checker/bin/javac -processor org.checkerframework.checker.nullness.NullnessChecker CF.java
```

returns the following false-positive

```
CF.java:[15,36] error: [dereference.of.nullable] dereference of possibly-null reference other.ref
```
(first else-if). `other.ref` shouldn't be `null` at this point because otherwise the first branch would have been taken (the field is final so it can't be re-assigned by any possible side-effects). I tested against checkerframework 3.17.0 and Java 8.

It's not a showstopper, since one can easily re-arrange the if statements to make it work, but I was a little bit suprised when this popped up during our code-analysis.

**Comment by jpschewe** [comment]
@mtf90  According to #5016 it seems that checker doesn't handle compound conditionals. In addition it's possible that another thread could modify the local class variable between statements. So you need to make 2 changes here, one is to make a local copy of the class variable reference to ensure that it doesn't change across the conditions and the second is to rework the conditionals to all be simple boolean conditions like the solution that @mernst  put in #5016.
